### PR TITLE
Free some locks earlier for queries.

### PR DESCRIPTION
**Pull request by cjgillot**
Keeping the locks held made `RUST_FORBID_DEP_GRAPH_EDGE` ICE.

### CODE REVIEW

**Comment by oli-obk** [comment]
Is there a self contained repro for the env var that we could add a test for?

**Comment by cjgillot** [comment]
@bors try @rust-timer queue

@oli-obk I'll see if I can make one.

File: compiler/rustc_query_system/src/query/caches.rs [hunk 1]
```diff
@@ -278,7 +278,9 @@ where
         let lock = self.cache.get_shard_by_hash(key.index() as u64).lock();
         #[cfg(not(parallel_compiler))]
         let lock = self.cache.lock();
-        if let Some(Some(value)) = lock.get(*key) {
+        if let Some(&Some(value)) = lock.get(*key) {
+            // Drop the lock before calling user-provided code.
+            drop(lock);
```
Comment (Zoxc) [inline compiler/rustc_query_system/src/query/caches.rs:283]: This kind of conditional drop could cause worse code generation.

**Comment by oli-obk** [comment]
@rustbot author

r=me with a test

**Comment by JohnCSimon** [comment]
@cjgillot  Ping from triage:  Can you post your status on this PR? This has sat idle for a few months.

**Comment by cjgillot** [comment]
Waiting for me to write the test.

**Pull request closed without merging**